extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/kernel/controller/ControllerQueryKernel.java

```diff
@@ -643,6 +644,20 @@ public void failStage(final StageId stageId)
     doWithStageTracker(stageId, ControllerStageTracker::fail);
   }
 
+  /**
+   * Returns the set of all worker numbers that have participated in work done so far by this query.
+   */
+  public IntSet getAllParticipatingWorkers()
```

Comment:
Can this be cached, by maintaining a set of the participating workers, and whenever any new worker starts/stops participating, updating that set? Not a big deal though.

Reply:
It's only called twice per query, so I think it's not worth the extra complexity to keep the cache up to date.